### PR TITLE
Improve v-select of task state filter in responsive

### LINKED ISSUE #513: Tree component displays task state filters in two lines, breaking the layout
**Describe the bug**

Depending on the screen size, you may get two lines in the select element if you have selected many task states.

image

One alternative could be to show only 4 or 5 items, and show `…` or another indicator that there are more items selected.

**Release version(s) and/or repository branch(es) affected?**

`master`

**Steps to reproduce the bug**

1. Select all the task states in the tree filter
2. Play with your browser width, resizing the window or using FFox mobile preview

**Expected behavior**

Concise layout, without the extra line (which most likely causes reflow, and may result in unexpected layout issues depending on other components/browser engine/etc)

**Screenshots**

image

**Additional context**

Related to #512 and #508 

**Pull requests welcome!**
This is an Open Source project - please consider contributing a bug fix
yourself (please read `CONTRIBUTING.md` before starting any work though).

### PR DESCRIPTION
**Pull request by SoldierCorp**
These changes close #513 

Evidence of the changes
Peek 2020-10-20 18-42

The limit is 4 but is hardcoded in the Tree.vue, so if you are planning to have this number as a limit in a config file, let me know if you already have one to place it there.

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).

- [x] Already covered by existing tests.

- [x] Appropriate change log entry included.

- [x] No documentation update required.

### CODE REVIEW

**Comment by kinow** [comment]
@SoldierCorp this is exactly what I had in mind! Will finish some work in another branch, and then will check out yours to review it. But from your GIF it seems to be pretty much done and working as expected! :tada:

**Review by kinow — COMMENTED** [review]
@SoldierCorp what do you think about moving the constant `4` to the component's `data`?

e.g.

```diff
diff --git a/src/components/cylc/tree/Tree.vue b/src/components/cylc/tree/Tree.vue
index 0ed4338..548250a 100644
--- a/src/components/cylc/tree/Tree.vue
+++ b/src/components/cylc/tree/Tree.vue
@@ -56,14 +56,14 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
             <span class="ml-2">{{ slotProps.item.value.toLowerCase() }}</span>
           </template>
           <template v-slot:selection="slotProps">
-            <div class="mr-2" v-if="slotProps.index >= 0 && slotProps.index < 4">
+            <div class="mr-2" v-if="slotProps.index >= 0 && slotProps.index < maximumTasks">
               <Task :status="slotProps.item.value.toLowerCase()" :progress=0 />
             </div>
             <span
               v-if="slotProps.index === 4"
               class="grey--text caption"
             >
-              (+{{ tasksFilter.states.length - 4 }})
+              (+{{ tasksFilter.states.length - maximumTasks }})
             </span>
           </template>
         </v-select>
@@ -138,7 +138,8 @@ export default {
         name: '',
         states: []
       },
-      activeFilters: null
+      activeFilters: null,
+      maximumTasks: 4
     }
   },
   computed: {
```

And maybe we can have an e2e test too?

e.g.

```diff
diff --git a/tests/e2e/specs/tree.js b/tests/e2e/specs/tree.js
index ea3a4bb..dcb3e33 100644
--- a/tests/e2e/specs/tree.js
+++ b/tests/e2e/specs/tree.js
@@ -15,6 +15,8 @@
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
 
+import TaskState from '@/model/TaskState.model'
+
 describe('Tree component', () => {
   it('Should display cycle points for the mocked workflow', () => {
     cy.visit('/#/workflows/one')
@@ -220,5 +222,23 @@ describe('Tree component', () => {
         .get('.node-data-task-proxy:visible')
         .should('have.length', 1)
     })
+    it('should show a summary of tasks if the number of selected items is greater than the maximum limit', () => {
+      cy.visit('/#/tree/one')
+      cy
+        .get('#c-tree-filter-task-states')
+        .click({ force: true })
+      // eslint-disable-next-line no-lone-blocks
+      TaskState.enumValues.forEach(state => {
+        cy
+          .get('.v-list-item')
+          .contains(state.name.toLowerCase())
+          .click({ force: true })
+      })
+      cy
+        .get('.v-select__slot')
+        .should($select => {
+          expect($select).to.contain('(+')
+        })
+    })
   })
 })
```

Bruno

File: src/components/cylc/tree/Tree.vue [hunk 1]
```diff
@@ -34,7 +36,9 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
         ></v-text-field>
       </v-col>
       <v-col
-        class="ma-0 pa-md-0 col-sm-12 col-md-4">
+        cols="12"
+        md="4"
+        class="ma-0 py-md-0 pa-md-0">
```
[the remark this answers is not in the document]
Comment (SoldierCorp) [reply src/components/cylc/tree/Tree.vue:41]: Yes, I got it from the Vuetify docs.

File: CHANGES.md [hunk 2]
```diff
@@ -44,6 +44,8 @@ by tasks name and tasks states.
 [#505](https://github.com/cylc/cylc-ui/pull/505) - Use cycle point ID
 in the tree view structure.
 
+[#522](https://github.com/cylc/cylc-ui/pull/522) - Improve v-select of task state filter in responsive
```
[the remark this answers is not in the document]
Comment (SoldierCorp) [reply CHANGES.md:47]: 🙂

**Comment by SoldierCorp** [comment]
> @SoldierCorp what do you think about moving the constant `4` to the component's `data`?
> 
> e.g.

Done!


> And maybe we can have an e2e test too?
> 
> e.g.

Sure thing!

I have a couple of questions because I am not so good with e2e tests, but I want to implement this for sure.

I already added the code you wrote down and when I ran `yarn test:e2e`, I got errors on every test even before adding the new one. Do I have to do something else before?

image

**Comment by kinow** [comment]
>I have a couple of questions because I am not so good with e2e tests, but I want to implement this for sure.
>
>I already added the code you wrote down and when I ran yarn test:e2e, I got errors on every test even before adding the new one. Do I have to do something else before?

@SoldierCorp sorry, I had planned to include a paragraph with instructions for running the e2e tests but forgot to do so.

You can try the following:

```bash
$ export NODE_ENV=offline
$ yarn run test:e2e
```

We [are changing](https://github.com/cylc/cylc-ui/pull/519) `package.json` to make `yarn run test:e2e` to automatically choose the offline environment.

This offline environment is used [to mock the service](https://github.com/cylc/cylc-ui/blob/d60c6255f1b0c96c280fed75f32f9b9c7ddc8769/vue.config.js#L70-L79) that fetches workflow data. This way you can run `yarn run serve` and get offline data, without the need of the backend. The e2e tests are built using this same mechanism.

**Comment by SoldierCorp** [comment]
Thank you @kinow, it's working now!

I added the test you shared and moved the constant to the component's data.

**Review by kinow — APPROVED** [review]
Tested locally, no browser console errors, UI working fine. Approving (@oliver-sanders / @hjoliver  anyone up for second reviewer?)

**Comment by SoldierCorp** [comment]
It was fun, thank you @kinow 😃

**Review by hjoliver — APPROVED** [review]
Looks great to me; thanks @SoldierCorp :+1:

**Pull request merged**